## 1. Importing Libraries

In [211]:
import os
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

## 2. Reading Data

In [222]:
#os.getcwd()

In [223]:
project_dir = r"D:\\SAURABH\\spring 2024\\ML_Projects\\AWS SageMaker FlightFarePredictor"
data_dir = 'data'

In [224]:
def get_data(name):
    file_name = f"{name}.csv"
    file_path = os.path.join(project_dir, data_dir, file_name)
    return pd.read_csv(file_path)

In [225]:
flights = get_data("flight_price")
flights
                   

Airline Date_of_Journey    Source Destination  \
0           IndiGo      24/03/2019  Banglore   New Delhi   
1        Air India       1/05/2019   Kolkata    Banglore   
2      Jet Airways       9/06/2019     Delhi      Cochin   
3           IndiGo      12/05/2019   Kolkata    Banglore   
4           IndiGo      01/03/2019  Banglore   New Delhi   
...            ...             ...       ...         ...   
10678     Air Asia       9/04/2019   Kolkata    Banglore   
10679    Air India      27/04/2019   Kolkata    Banglore   
10680  Jet Airways      27/04/2019  Banglore       Delhi   
10681      Vistara      01/03/2019  Banglore   New Delhi   
10682    Air India       9/05/2019     Delhi      Cochin   

                       Route Dep_Time  Arrival_Time Duration Total_Stops  \
0                  BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1      CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2      DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h     2 stops   
3            CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4            BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   
...                      ...      ...           ...      ...         ...   
10678              CCU → BLR    19:55         22:25   2h 30m    non-stop   
10679              CCU → BLR    20:45         23:20   2h 35m    non-stop   
10680              BLR → DEL    08:20         11:20       3h    non-stop   
10681              BLR → DEL    11:30         14:10   2h 40m    non-stop   
10682  DEL → GOI → BOM → COK    10:55         19:15   8h 20m     2 stops   

      Additional_Info  Price  
0             No info   3897  
1             No info   7662  
2             No info  13882  
3             No info   6218  
4             No info  13302  
...               ...    ...  
10678         No info   4107  
10679         No info   4145  
10680         No info   7229  
10681         No info  12648  
10682         No info  11753  

[10683 rows x 11 columns]

In [226]:
# flights = pd.read_csv(r"../data/flight_price.csv")

In [227]:
# The info method provides additional information about the flights dataset.
flights.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


- **Observations:**
  - **RangeIndex:** The dataset contains 10683 rows, and 11 features.
  - **Total_Stops, Route:** There is 1 missing value in the 'Total_Stops' and 'Route' column, which might be present in the same or different rows. This needs further consideration.
  - Some features data types are not appropriate


## 3. Preliminary Analysis

### 3.1 Check Data types

In [228]:
flights.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

In [229]:
flights.dtypes

Airline            object
Date_of_Journey    object
Source             object
Destination        object
Route              object
Dep_Time           object
Arrival_Time       object
Duration           object
Total_Stops        object
Additional_Info    object
Price               int64
dtype: object

### 3.2 Check for Duplicates

In [230]:
flights.duplicated().sum()

220

In [231]:
# loc to filter

(     
    flights
    .loc[flights .duplicated(keep = False)]
    .sort_values(["Airline","Date_of_Journey","Source","Route"])
)

Airline Date_of_Journey    Source Destination                  Route  \
6321  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
9848  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
572   Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
8168  Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
1495  Air India       1/04/2019   Kolkata    Banglore  CCU → DEL → COK → BLR   
...         ...             ...       ...         ...                    ...   
2692   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2870   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
3711   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2634    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   
5470    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   

     Dep_Time  Arrival_Time Duration Total_Stops  \
6321    08:50  23:55 02 Mar   39h 5m     2 stops   
9848    08:50  23:55 02 Mar   39h 5m     2 stops   
572     21:10         23:55   2h 45m    non-stop   
8168    21:10         23:55   2h 45m    non-stop   
1495    10:00  01:20 02 Apr  15h 20m     2 stops   
...       ...           ...      ...         ...   
2692    05:45         08:35   2h 50m    non-stop   
2870    05:45         08:35   2h 50m    non-stop   
3711    20:30         23:20   2h 50m    non-stop   
2634    11:30         14:10   2h 40m    non-stop   
5470    11:30         14:10   2h 40m    non-stop   

                   Additional_Info  Price  
6321                       No info  17135  
9848                       No info  17135  
572                        No info   7591  
8168                       No info   7591  
1495                       No info  10408  
...                            ...    ...  
2692  No check-in baggage included   4273  
2870  No check-in baggage included   4273  
3711  No check-in baggage included   3873  
2634                       No info   5403  
5470                       No info   5403  

[415 rows x 11 columns]

### 3.3 Observations

- The type of `Date_of_Journey`, `Dep_Time` and `Arrival_Time` should be changed to datetime
- The type of `Duration` and `Total_Stops` is mixed. It should be numeric type.
- There're 220 duplicates. These should be removed.




## 4. Detailed Analysis (wrt each feature)

### Feature 1: Airline

In [232]:
flights.Airline

0             IndiGo
1          Air India
2        Jet Airways
3             IndiGo
4             IndiGo
            ...     
10678       Air Asia
10679      Air India
10680    Jet Airways
10681        Vistara
10682      Air India
Name: Airline, Length: 10683, dtype: object

In [233]:
# when cleaning a categorical feature we can use unique methods 
# after which we can analyze if there are any discrepancies 

flights.Airline.unique()

array(['IndiGo', 'Air India', 'Jet Airways', 'SpiceJet',
       'Multiple carriers', 'GoAir', 'Vistara', 'Air Asia',
       'Vistara Premium economy', 'Jet Airways Business',
       'Multiple carriers Premium economy', 'Trujet'], dtype=object)

- Some of the entries have inaccurate values 
- e.g. Vistara and Vistara Premium economy(just a class of vistara airlines) so its the same airlines we can replace that with just airline name

In [234]:
# including the code under paranthesis to improve readibility of the code
# so here we did experimentation once we are satisfied with the result will write same code inside function
(
    flights
    .Airline
    .str.replace(" Premium economy","")
    .str.replace(" Business","")
    .str.title() # make string first letter capital
    .unique()
)

     

array(['Indigo', 'Air India', 'Jet Airways', 'Spicejet',
       'Multiple Carriers', 'Goair', 'Vistara', 'Air Asia', 'Trujet'],
      dtype=object)

### Date of Journey

In [235]:
flights.Date_of_Journey

0        24/03/2019
1         1/05/2019
2         9/06/2019
3        12/05/2019
4        01/03/2019
            ...    
10678     9/04/2019
10679    27/04/2019
10680    27/04/2019
10681    01/03/2019
10682     9/05/2019
Name: Date_of_Journey, Length: 10683, dtype: object

- dtype is object will convert it into date time feature

In [236]:
pd.to_datetime(flights.Date_of_Journey, dayfirst=True)

0       2019-03-24
1       2019-05-01
2       2019-06-09
3       2019-05-12
4       2019-03-01
           ...    
10678   2019-04-09
10679   2019-04-27
10680   2019-04-27
10681   2019-03-01
10682   2019-05-09
Name: Date_of_Journey, Length: 10683, dtype: datetime64[ns]

- now the dtype is datetime

### Source

In [237]:
flights.Source.unique()

array(['Banglore', 'Kolkata', 'Delhi', 'Chennai', 'Mumbai'], dtype=object)

### Destination

In [238]:
flights.Destination.unique()

array(['New Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Delhi', 'Hyderabad'],
      dtype=object)

- there seems to be no inconsistencies wrt source and destination columns hence no further cleaning required

### Route

In [239]:
# flights.Route.unique()
# We will drop the 'Route' column because it is redundant. It does not provide any additional information beyond what is already given by 'Source', 'Destination', and the number of stops.

### Dep_Time

In [240]:
flights.Dep_Time

0        22:20
1        05:50
2        09:25
3        18:05
4        16:50
         ...  
10678    19:55
10679    20:45
10680    08:20
10681    11:30
10682    10:55
Name: Dep_Time, Length: 10683, dtype: object

- We will filter out any values that are not in the format of numbers:numbers.


In [241]:
(
    flights
    .Dep_Time
    .loc[lambda ser: ser.str.contains("[^0-9:]")] 
    # This code filters the 'Dep_Time' column in the 'flights' DataFrame.
    # The 'loc' method is used to filter the series, selecting only the values that contain characters
    # other than numbers and colons (':'). The lambda function passed to 'loc' applies a string
    # method 'str.contains' with a regular expression '[^0-9:]' which matches any character that is not
    # a digit or a colon. This helps in identifying and filtering out any departure times that are not 
    # in the correct 'numbers:numbers' format.
)


Series([], Name: Dep_Time, dtype: object)

- result is a empty series i.e all the rows have only digits and columns so there are no incorrect format

In [242]:
pd.to_datetime(flights.Dep_Time).dt.time
# lets implement the same in our function

0        22:20:00
1        05:50:00
2        09:25:00
3        18:05:00
4        16:50:00
           ...   
10678    19:55:00
10679    20:45:00
10680    08:20:00
10681    11:30:00
10682    10:55:00
Name: Dep_Time, Length: 10683, dtype: object

### Arrival_Time

In [243]:
flights.Arrival_Time

0        01:10 22 Mar
1               13:15
2        04:25 10 Jun
3               23:30
4               21:35
             ...     
10678           22:25
10679           23:20
10680           11:20
10681           14:10
10682           19:15
Name: Arrival_Time, Length: 10683, dtype: object

In [244]:
(
    flights
    .Arrival_Time
    .loc[lambda ser: ser.str.contains("[^0-9:]")] 
    .str.split(" ", n=1) # we are specifying to make a split wherever there is space but n=1 i.e only on first space
    .str.get(1)
    .unique()
    
    
    
)


array(['22 Mar', '10 Jun', '13 Mar', '02 Mar', '10 May', '04 Mar',
       '13 Jun', '28 May', '19 Mar', '07 May', '02 Jun', '16 Jun',
       '19 May', '16 May', '28 Jun', '02 May', '28 Mar', '19 Jun',
       '04 Apr', '25 Mar', '07 Mar', '25 Jun', '07 Jun', '25 May',
       '13 May', '16 Mar', '22 May', '10 Apr', '04 Jun', '20 May',
       '28 Apr', '25 Apr', '10 Mar', '19 Apr', '13 Apr', '02 Apr',
       '23 Mar', '22 Apr', '11 May', '07 Apr', '03 May', '08 Mar',
       '03 Mar', '05 Mar', '22 Jun', '04 May', '26 May', '16 Apr',
       '26 Jun', '29 May', '29 Jun', '29 Mar', '23 May', '17 Jun'],
      dtype=object)

- There are 4335 rows which contains some other characters
- if we see output its just date there are no special characters so its clean the values are valid

### Duration

In [245]:
flights.Duration

0        2h 50m
1        7h 25m
2           19h
3        5h 25m
4        4h 45m
          ...  
10678    2h 30m
10679    2h 35m
10680        3h
10681    2h 40m
10682    8h 20m
Name: Duration, Length: 10683, dtype: object

- will convert row values into minutes

In [246]:
(
    flights
    .Duration
    .loc[lambda ser: ~ser.str.contains("m")] # give me rows that do not conatain m
    .unique()
)

array(['19h', '23h', '22h', '12h', '3h', '5h', '10h', '18h', '24h', '15h',
       '16h', '8h', '14h', '20h', '13h', '11h', '9h', '27h', '26h', '4h',
       '7h', '30h', '21h', '28h', '47h', '6h', '25h', '38h', '34h'],
      dtype=object)

In [247]:
(
    flights
    .Duration
    .loc[lambda ser: ~ser.str.contains("h")] # give me rows that do not conatain m
    #.unique()
)

6474    5m
Name: Duration, dtype: object

- theres one flight which is only 5 minutes doesnt make sense

In [248]:
flights.iloc[[6474]]

Airline Date_of_Journey  Source Destination                  Route  \
6474  Air India       6/03/2019  Mumbai   Hyderabad  BOM → GOI → PNQ → HYD   

     Dep_Time Arrival_Time Duration Total_Stops Additional_Info  Price  
6474    16:50        16:55       5m     2 stops         No info  17327

- So the observation index no 6474 has the duration of 5 minutes. This is clearly wrong. Will delete this observation.

In [249]:
(
    flights
    .Duration
    .drop(index=6474)
    .str.split(" ", expand=True)  # By setting expand=True, the split values will be expanded into separate columns in a DataFrame. Without expand=True, the result would be a Series.
    .set_axis(["hour","minute"], axis=1)
    .assign(
        hour=lambda df2: (
            df2
            .hour
            .str.replace("h"," ")
            .astype(int)
            .mul(60)
        ),
        minute=lambda df2: (
            df2
            .minute
            .str.replace("m"," ")
            .fillna("0")
            .astype(int)
        )
            
    )
    #.dtypes
    #.isna().sum()
    .sum(axis=1)
)


0         170
1         445
2        1140
3         325
4         285
         ... 
10678     150
10679     155
10680     180
10681     160
10682     500
Length: 10682, dtype: int64

- sucessfully converted every value into minute
- dtype from object to int

In [250]:
# to validate will se both columns simultaneously

(
    flights
    .Duration
    .drop(index=6474)
    .str.split(" ", expand=True)  # By setting expand=True, the split values will be expanded into separate columns in a DataFrame. Without expand=True, the result would be a Series.
    .set_axis(["hour","minute"], axis=1)
    .assign(
        hour=lambda df2: (
            df2
            .hour
            .str.replace("h"," ")
            .astype(int)
            .mul(60)
        ),
        minute=lambda df2: (
            df2
            .minute
            .str.replace("m"," ")
            .fillna("0")
            .astype(int)
        )
            
    )
    #.dtypes
    #.isna().sum()
    .sum(axis=1)
    .rename("duration_minutes")
    .to_frame()
    .join(flights.Duration.drop(index=6474))
)


duration_minutes Duration
0                   170   2h 50m
1                   445   7h 25m
2                  1140      19h
3                   325   5h 25m
4                   285   4h 45m
...                 ...      ...
10678               150   2h 30m
10679               155   2h 35m
10680               180       3h
10681               160   2h 40m
10682               500   8h 20m

[10682 rows x 2 columns]

### Total_Stops

In [251]:
flights.Total_Stops.unique()

array(['non-stop', '2 stops', '1 stop', '3 stops', nan, '4 stops'],
      dtype=object)

- nan - there are some missing values we will not analyze missing values here will see it in EDA step

In [252]:
(
    flights
    .Total_Stops
    .replace("non-stop", "0")  # Replacing "non-stop" with "0"
    .str.replace(" stops?", "", regex=True)  # Removing the word "stop" or "stops"
    #.unique()  
    #.astype(int)  # astype will not work since in pandas, missing values (NaN) are float, so we cannot convert them into integers. Instead, we will apply pipe to convert the current state of the series to numeric, which will result in floats.
    .pipe(lambda ser: pd.to_numeric(ser))
    
)


0        0.0
1        2.0
2        2.0
3        1.0
4        1.0
        ... 
10678    0.0
10679    0.0
10680    0.0
10681    0.0
10682    2.0
Name: Total_Stops, Length: 10683, dtype: float64

### Additional_Info

In [253]:
flights.Additional_Info.unique()

array(['No info', 'In-flight meal not included',
       'No check-in baggage included', '1 Short layover', 'No Info',
       '1 Long layover', 'Change airports', 'Business class',
       'Red-eye flight', '2 Long layover'], dtype=object)

- will convert no info to No Info

## 5 Cleaning Operations

In [254]:
def convert_to_minutes(ser):
    return (
        ser
        .str.split(" ", expand=True)  # By setting expand=True, the split values will be expanded into separate columns in a DataFrame. Without expand=True, the result would be a Series.
        .set_axis(["hour","minute"], axis=1)
        .assign(
            hour=lambda df2: (
                df2
                .hour
                .str.replace("h"," ")
                .astype(int)
                .mul(60)
            ),
            minute=lambda df2: (
                df2
                .minute
                .str.replace("m"," ")
                .fillna("0")
                .astype(int)
            )

        )
        #.dtypes
        #.isna().sum()
        .sum(axis=1)
    )

In [255]:
# functional programming helps to avoid overwriting the data
# if i dont want any cleaning method i can just comment it down without affecting my original data

def clean_data(df):
    return (
        df
        .drop(index=[6474])
        .drop_duplicates()
        .assign(**{                                            # dictionary unpacking in python
            col: df[col].str.strip()
            for col in df.select_dtypes(include = "O").columns  # strip method not supportable for int data type
        })
        .rename(columns=str.lower)           
        .assign(
            airline = lambda df2: (
                df2
                .airline
                .str.replace(" Premium economy","")
                .str.replace(" Business","")
                .str.title() # make string first letter capital
            ), # inside assign method if we want to update multiple feature will use comma
            date_of_journey=lambda df2: pd.to_datetime(df2.date_of_journey, dayfirst=True),
            dep_time=lambda df2: pd.to_datetime(df2.dep_time).dt.time,
            arrival_time=lambda df2: pd.to_datetime(df2.arrival_time).dt.time,
            #duration=lambda df2: convert_to_minutes(df2.duration)   # this will work 
            duration=lambda df2: df2.duration.pipe(convert_to_minutes), #good practice # pipe method can be called on dataframe or series 
            total_stops=lambda df2: (
                df2
                .total_stops
                .replace("non-stop", "0")
                .str.replace(" stops?", "", regex=True)
                .pipe(lambda ser: pd.to_numeric(ser))
            ),
            additional_info=lambda df2: (
                df2
                .additional_info
                .replace("No info","No Info")
            )

        )
        .drop(columns='route')
    )
                
            
                
    
            
  
            


In [256]:
flights_cleaned_data = clean_data(flights)
flights_cleaned_data

airline date_of_journey    source destination  dep_time  \
0           Indigo      2019-03-24  Banglore   New Delhi  22:20:00   
1        Air India      2019-05-01   Kolkata    Banglore  05:50:00   
2      Jet Airways      2019-06-09     Delhi      Cochin  09:25:00   
3           Indigo      2019-05-12   Kolkata    Banglore  18:05:00   
4           Indigo      2019-03-01  Banglore   New Delhi  16:50:00   
...            ...             ...       ...         ...       ...   
10678     Air Asia      2019-04-09   Kolkata    Banglore  19:55:00   
10679    Air India      2019-04-27   Kolkata    Banglore  20:45:00   
10680  Jet Airways      2019-04-27  Banglore       Delhi  08:20:00   
10681      Vistara      2019-03-01  Banglore   New Delhi  11:30:00   
10682    Air India      2019-05-09     Delhi      Cochin  10:55:00   

      arrival_time  duration  total_stops additional_info  price  
0         01:10:00       170          0.0         No Info   3897  
1         13:15:00       445          2.0         No Info   7662  
2         04:25:00      1140          2.0         No Info  13882  
3         23:30:00       325          1.0         No Info   6218  
4         21:35:00       285          1.0         No Info  13302  
...            ...       ...          ...             ...    ...  
10678     22:25:00       150          0.0         No Info   4107  
10679     23:20:00       155          0.0         No Info   4145  
10680     11:20:00       180          0.0         No Info   7229  
10681     14:10:00       160          0.0         No Info  12648  
10682     19:15:00       500          2.0         No Info  11753  

[10462 rows x 10 columns]

In [257]:
#clean_data(flights).airline.unique()

In [258]:
# clean_data(flights).dtypes

## 6. Splitting Data

In [259]:
flights_final_data = flights_cleaned_data


In [260]:
X = flights_final_data.drop(columns="price")
y = flights_final_data.price.copy()

In [261]:
#X

In [262]:
#y

In [263]:
X_, X_test, y_, y_test = train_test_split(X, y, test_size=0.2, random_state=17)
# using X_ and y_ will make our train and validation set
X_train, X_val, y_train, y_val = train_test_split(X_, y_, test_size=0.2, random_state=17)

print(X_train.shape, y_train.shape)
print(X_val.shape, y_val.shape)
print(X_test.shape, y_test.shape)

(6695, 9) (6695,)
(1674, 9) (1674,)
(2093, 9) (2093,)


## Export the Subsets

In [264]:

def export_data(X, y, name):
    file_name = f"{name}.csv"
    file_path = os.path.join(project_dir, data_dir, file_name)

    X.join(y).to_csv(file_path, index=False)

    return pd.read_csv(file_path).head()

In [265]:
export_data(X_train, y_train, "train_set")

airline date_of_journey   source destination  dep_time arrival_time  \
0       Indigo      2019-06-06  Kolkata    Banglore  20:25:00     23:10:00   
1    Air India      2019-05-21    Delhi      Cochin  17:15:00     19:15:00   
2  Jet Airways      2019-03-24  Kolkata    Banglore  21:10:00     16:20:00   
3       Indigo      2019-04-01    Delhi      Cochin  14:20:00     17:35:00   
4  Jet Airways      2019-03-15   Mumbai   Hyderabad  10:20:00     11:50:00   

   duration  total_stops              additional_info  price  
0       165          0.0                      No Info   4804  
1      1560          2.0                      No Info  11989  
2      1150          1.0  In-flight meal not included  10031  
3       195          0.0                      No Info   4729  
4        90          0.0                      No Info   8040

In [266]:
export_data(X_val, y_val, "val_set")


airline date_of_journey   source destination  dep_time arrival_time  \
0  Jet Airways      2019-06-12    Delhi      Cochin  05:30:00     04:25:00   
1  Jet Airways      2019-06-09  Kolkata    Banglore  20:25:00     23:35:00   
2       Indigo      2019-06-03  Chennai     Kolkata  22:05:00     00:25:00   
3  Jet Airways      2019-03-24  Kolkata    Banglore  08:25:00     22:35:00   
4      Vistara      2019-06-15  Chennai     Kolkata  17:45:00     20:05:00   

   duration  total_stops              additional_info  price  
0      1375          2.0  In-flight meal not included  10368  
1      1630          1.0                      No Info  12692  
2       140          0.0                      No Info   5277  
3       850          1.0                      No Info  13759  
4       140          0.0                      No Info  11982

In [267]:
export_data(X_test, y_test, "test_set")


airline date_of_journey    source destination  dep_time  \
0        Jet Airways      2019-03-01  Banglore   New Delhi  08:00:00   
1             Indigo      2019-06-15  Banglore       Delhi  04:00:00   
2  Multiple Carriers      2019-06-12     Delhi      Cochin  07:30:00   
3          Air India      2019-04-21  Banglore       Delhi  06:10:00   
4  Multiple Carriers      2019-06-03     Delhi      Cochin  15:00:00   

  arrival_time  duration  total_stops additional_info  price  
0     15:15:00       435          1.0         No Info  22270  
1     06:50:00       170          0.0         No Info   4423  
2     19:00:00       690          1.0         No Info  10877  
3     08:55:00       165          0.0         No Info   6121  
4     01:30:00       630          1.0         No Info  13377